aws_signing_helper/pkcs11_signer.go

```
@@ -0,0 +1,214 @@
+	}
+	defer p.CloseSession(session)
+
+	var templatecrt []*pkcs11.Attribute
+
+
+	templatecrt = append(templatecrt, pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_CERTIFICATE))
+
+	if err = p.FindObjectsInit(session, templatecrt); err != nil {
+		fmt.Println("nil, err")
+	}
+	defer func() {
+		finalErr := p.FindObjectsFinal(session)
+		if err == nil {
+			err = finalErr
+		}
+	}()
+
+	objs, b, err := p.FindObjects(session, 1)
+	for err == nil {
+		var o []pkcs11.ObjectHandle
+		o, b, err = p.FindObjects(session, 1)
+		if err != nil {
+			continue
+		}
+		if len(o) == 0 {
+			break
+		}
+		objs = append(objs, o...)
+	}
+	if err != nil {
+		log.Println("Failed to find: ", b)
+		if len(objs) == 0 {
+			fmt.Println("nil")
+		}
+	}
+
+	_ = p.FindObjectsFinal(session)  // found nothing
+
+	for i := range objs {
+		attributescrt := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_VALUE, 0),
+		}
+
+		if attributescrt, err = p.GetAttributeValue(session, objs[i], attributescrt); err != nil {
+			fmt.Println("nil, err2")
+		}
+
+		rawCertificate := attributescrt[0].Value
+		cert, err := x509.ParseCertificate(rawCertificate)
+		if err != nil {fmt.Println("Error parsing certificate")}
+		serialNumber := cert.SerialNumber.String()
+		certLocated.Subject = cert.Subject.CommonName 
+		certLocated.Issuer = cert.Issuer.CommonName
+		certLocated.SerialNumber, _ = new(big.Int).SetString(serialNumber, 16)
+		if reflect.DeepEqual(certLocated, certIdentifier) {
+			return i, cert, nil
+		} else {
+		}
+	}
+
+
+	return 1, nil, errors.New("unsupported certificate")
+}
+
+func (pkcs11Signer PKCS11Signer) Public() crypto.PublicKey {
+	return nil
+}
+
+func (pkcs11Signer PKCS11Signer) Close() {
+}
+
+func (signer PKCS11Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+		lib := signer.libpkcs11
+		pin := signer.pinpkcs11
+		id := signer.idpkcs11
+
+		p := pkcs11.New(lib) 
+		err = p.Initialize()
+		if err != nil {
+			panic(err)
```

Minor, but to stay consistent, it would be better to throw an error here. The caller of the function can choose what to do with the error.

---

Returning the error that can then be used by the caller.